File: core/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.java
package com.dbc;

import java.util.HashMap;
import java.util.Map;

public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap();
        int left = 0, right = 0, res = 0;
        while (right < s.length()) {
            map.put(s.charAt(right), map.getOrDefault(s.charAt(right), 0) + 1);
            while (map.get(s.charAt(right)) >= 2) {
                map.put(s.charAt(left), map.get(s.charAt(left)) - 1);
                left++;
            }
            res = Math.max(res, right - left + 1);
            right++;
        }
        return res;
    }
}

File: core/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.py
from collections import Counter

def lengthOfLongestSubstring(s: str) -> int:
    counter = Counter()
    left, right, length = 0, 0, len(s)
    res = 0
    while right < length:
        counter[s[right]] += 1
        while counter[s[right]] >= 2:
            counter[s[left]] -= 1
            left += 1

        res = max(res, right - left + 1)
        right += 1

    return res



